drv/stm32h7-sprot-server/src/main.rs

```diff
@@ -1363,6 +1363,172 @@ impl<S: SpiServer> idl::InOrderSpRotImpl for ServerImpl<S> {
+            Err(e) => Err(AttestOrSprotError::Sprot(e).into()),
+        }
+    }
+
+    fn tq_cert_chain_len(
+        &mut self,
+        _: &userlib::RecvMessage,
+    ) -> Result<u32, idol_runtime::RequestError<AttestOrSprotError>> {
+        let body = ReqBody::Attest(AttestReq::TqCertChainLen);
+        let tx_size = Request::pack(&body, self.tx_buf);
+        let rsp = self.do_send_recv_retries(tx_size, TIMEOUT_QUICK, 1)?;
+        match rsp.body {
+            Ok(RspBody::Attest(Ok(AttestRsp::TqCertChainLen(s)))) => Ok(s),
+            Ok(RspBody::Attest(Err(e))) => {
+                Err(AttestOrSprotError::Attest(e).into())
+            }
+            Ok(RspBody::Attest(_)) | Ok(_) => Err(AttestOrSprotError::Sprot(
+                SprotError::Protocol(SprotProtocolError::UnexpectedResponse),
+            )
+            .into()),
+            Err(e) => Err(AttestOrSprotError::Sprot(e).into()),
+        }
+    }
+
+    fn tq_cert_len(
+        &mut self,
+        _: &userlib::RecvMessage,
+        index: u32,
+    ) -> Result<u32, idol_runtime::RequestError<AttestOrSprotError>> {
+        let body = ReqBody::Attest(AttestReq::TqCertLen(index));
+        let tx_size = Request::pack(&body, self.tx_buf);
+        let rsp = self.do_send_recv_retries(tx_size, TIMEOUT_QUICK, 1)?;
+        match rsp.body {
+            Ok(RspBody::Attest(Ok(AttestRsp::TqCertLen(s)))) => Ok(s),
+            Ok(RspBody::Attest(Err(e))) => {
+                Err(AttestOrSprotError::Attest(e).into())
+            }
+            Ok(RspBody::Attest(_)) | Ok(_) => Err(AttestOrSprotError::Sprot(
+                SprotError::Protocol(SprotProtocolError::UnexpectedResponse),
+            )
+            .into()),
+            Err(e) => Err(AttestOrSprotError::Sprot(e).into()),
+        }
+    }
+
+    fn tq_sign(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        hash: idol_runtime::LenLimit<
+            idol_runtime::Leased<idol_runtime::R, [u8]>,
+            TQ_HASH_SIZE,
+        >,
+        dest: idol_runtime::Leased<idol_runtime::W, [u8]>,
+    ) -> Result<(), idol_runtime::RequestError<AttestOrSprotError>>
+    where
+        AttestOrSprotError: From<idol_runtime::ServerDeath>,
+    {
+        if hash.len() != TQ_HASH_SIZE {
+            return Err(
+                AttestOrSprotError::Attest(AttestError::BadLease).into()
+            );
+        }
+
+        let write_size = u32::try_from(dest.len()).unwrap_lite();
+
+        let body = ReqBody::Attest(AttestReq::TqSign {
+            write_size,
+        });
+        let tx_size = Request::pack_with_cb(&body, self.tx_buf, |buf| {
+            hash
+                .read_range(0..hash.len(), buf)
+                .map_err(|_| SprotProtocolError::TaskRestarted)?;
+            Ok::<usize, idol_runtime::RequestError<AttestOrSprotError>>(
+                hash.len(),
+            )
+        })?;
+
+        let rsp = self.do_send_recv_retries(tx_size, TIMEOUT_LONG, 1)?;
+
+        match rsp.body {
```

Comment:
**thought:** I didn't read and analyse the various method's code for line by line equivalence but this match expression is quite repetitive over all of these. I wonder if it couldn't be refactored into a common helper or two?